explainer.md

```diff
@@ -88,6 +89,21 @@ Items that are designed to be purchased once and last permanently in the user’
 itemService.acknowledge(purchaseToken, 'onetime');
 ```
 
+### Checking existing purchases
+
+The `listPurchases` method allows a client to get a list of items that are currently owned or purchased by the user. This may be necessary to check for entitlements (eg. whether a subscription, promotional code, or permanent upgrade is active) or to recover from network interruptions during a purchase (e.g. item is purchased but not yet acknowledged).
+
+
+```js
+purchases = await itemService.listPurchases();
+for (p in purchases) {
+  if (p.acknowledged)
+    DoSomethingWithEntitlement(p.itemId);
+  else
+    AcknowledgeAndDoSomething(p.itemId, p.token);
```

Comment:
Can we replace this with

```js
if (p.acknowledged)
  itemService.acknowledge(p.token);

DoSomethingWithEntitlement(p.itemId);
```

So we're actually showing how this interacts with the `acknowledge` method.

Reply:
You mean if (!p.acknowledged) ?

Reply:
Yeah. Updated.

Reply:
I actually considered it but I didn't want to make the code snippet too verbose, nor give an example doing the wrong thing. If we're going to make it more realistic there are more factors to consider:

- you'd likely want to record something in your own state when acknowledging
- you need to decide if it's "repeatable" or "onetime"
- you probably don't want to give the entitlement before acknowledge actually completes (acknowledge() is async)
- you probably don't want to await the acknowledgement while processing other entitlements

So it'd be more like:
```
if (p.acknowledged) {
  DoSomethingWithEntitlement(p.itemId);
} else {
  async (p) => {
    await itemService.acknowledge(p.token, OnetimeOrRepeatable(p.itemId));
    RecordSuccessfulPurchase(p);
    DoSomethingWithEntitlement(p.itemId);
  }(p);
}
```
I thought "Acknowledge" in a fake method name was clear enough without the detail. WDYT?

Reply:
I'd like to push back (sorry I took 6 days to get to this).

> I thought "Acknowledge" in a fake method name was clear enough without the detail.

We shouldn't hide calls to the API we're documenting inside an unseen method. Then we aren't showing how to use it. Having a slightly larger example is better if we show explicit usage of the API.

If there's extra complexity (like having to tell `acknowledge` whether it's one-time or repeatable), then a) that might reflect poorly on the design, but b) it's best to expose that issue rather than pretending it doesn't exist by not showing the steps you have to take.

So please do use the above code snippet instead.

> you probably don't want to await the acknowledgement while processing other entitlements

"Maybe" but I don't think it's necessary. For the purpose of example, it's fine to have it blocking, which simplifies the code (no need for an inner `async` block, and you can factor the `DoSomethingWithEntitlement` out of the if statement).

So it reduces to:

```js
if (!p.acknowledged) {
  await itemService.acknowledge(p.token, OnetimeOrRepeatable(p.itemId));
  RecordSuccessfulPurchase(p);
}

DoSomethingWithEntitlement(p.itemId);
```

Reply:
Used the example code you suggested. [Done]